### DISCUSSION TITLE
Byte Stream

### DISCUSSION THREAD

**Discussion by ajibrin** [FTHuld/neware_reader#19] [Q&A]
Hello, 

I am relatively new to coding and have what may be a simple question. I have a number of different models of neware devices that I pull data from. Some of these are high amp systems with only one current range and some are lower amp units with 3 current ranges. Each of these ranges needs a different current scaling factor to process the data correctly. My question is, how did you find out what parts of the byte stream correspond to certain values, like current, voltage, etc? Specifically, how do you know what part is the defining the current range so you can set the scaling factor appropriately?

**Comment by FTHuld** [comment]
Hi @ajibrin. The short answer is that we used trial and error to figure out how the nda file is structured. Specifically, I used HxD (free software) to open the file, then messed around with the bytes until I noticed a pattern. Once I had that I could figure out which part of the byte stream matched the bts software output (i.e. how wide the byte stream of each column is, type, etc). The python code pretty much just establishes the size of the header and iterates over the remaining file with the information I derived.

I believe `byte_stream[77:81]` will give you information about the scaling.
If you set beta=True in your options when converting the nda file, the code will try to convert the current to the correct values for you.
